system/jlib/jmutex.hpp

```diff
@@ -586,71 +586,120 @@ class jlib_decl Monitor: public Mutex
-                                }
-                                else
-                                {
-                                    writeWaiting++;
-                                    cs.leave();
-                                    if (timed)
-                                    {
-                                        if (!writeSem.wait(timeout)) {
-                                            cs.enter(); 
-                                            if (!writeSem.wait(0)) {
-                                                writeWaiting--;
-                                                cs.leave();
-                                                return false;
-                                            }
-                                            cs.leave();
-                                        }
-                                    }
-                                    else
-                                        writeSem.wait();
-                                    //NB: waiting and locks adjusted before the signal occurs.
-                                }
+    bool lockRead(bool timed, unsigned timeout)
+    {  
+        cs.enter(); 
+        if (writeLocks == 0) 
+        {
+            readLocks++;
+            cs.leave();
+        }
+        else
+        {
+            readWaiting++;
+            cs.leave();
+            if (timed)
+            {
+                if (!readSem.wait(timeout))
+                {
+                    cs.enter(); 
+                    if (!readSem.wait(0))
+                    {
+                        readWaiting--;
+                        cs.leave();
+                        return false;
+                    }
+                    cs.leave();
+                }
+            }
+            else
+                readSem.wait();
+            //NB: waiting and locks adjusted before the signal occurs.
+        }
+        return true;
+    }
+    bool lockWrite(bool timed, unsigned timeout)
+    {
+        cs.enter();
+        if ((readLocks == 0) && (writeLocks == 0))
+        {
+            writeLocks++;
+            cs.leave();
+        }
+        else
+        {
+            writeWaiting++;
+            cs.leave();
+            if (timed)
+            {
+                if (!writeSem.wait(timeout))
+                {
+                    cs.enter(); 
+                    if (!writeSem.wait(0))
+                    {
+                        writeWaiting--;
+                        cs.leave();
+                        return false;
+                    }
+                    cs.leave();
+                }
+            }
+            else
+                writeSem.wait();
+            //NB: waiting and locks adjusted before the signal occurs.
+        }
+#ifdef _DEBUG
+        exclWriteOwner = GetCurrentThreadId();
+#endif
+        return true;
+    }
+    bool changeToWrite(bool timed, unsigned timeout)
+    {
+        cs.enter();
+        if (writeLocks || (0 == readLocks))
```

Comment:
more: should this  throw an internal error instead of returning false?  ((readLocks==0) could be checked before the cs.

Reply:
or is the intention that it is ok to call this if it already has a write lock.  In which case the comment on the return is misleading.

Reply:
it's supposed to fail if already has a write lock, but a soft failure (return false) as opposed to an exception - in keeping with pre-exisitng timed lock read / write calls that return false on failure to get lock.

Reply:
>((readLocks==0) could be checked before the cs.

should it though ? it's an edge case, not generally expected, and I think code clearer when checked inside the crit.

Reply:
The comment implies it is not legal to call it when not read locked.  If that is kind-of allowed the comment should be changed.  E.g.
//if called when not readlocked then fail immediately.